### PR TITLE
add comma encoding and tests

### LINKED ISSUE #1785: Encode csv separator character in attribute value
If the csv separator character is found in an attribute value, it should be encoded

### PR DESCRIPTION
**Pull request by IvanHdzC**
Fixes #1785

### CODE REVIEW

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /**
  * Copyright 2014-2017 Telefonica Investigación y Desarrollo, S.A.U
- *
+ *purposes
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: The addition of this line seems to be a typo...

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /**
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: CNR entry about the changes in this PR should be added.

Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: CNR entry about the changes in this PR should be added.

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 3]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
+        String mappedServicePath = "newPath";
+        Map<String, String> headers = new HashMap<>();
+        headers.put(NGSIConstants.FLUME_HEADER_TIMESTAMP, timestamp);
+        headers.put(CommonConstants.HEADER_CORRELATOR_ID, correlatorId);
+        headers.put(NGSIConstants.FLUME_HEADER_TRANSACTION_ID, transactionId);
+        headers.put(CommonConstants.HEADER_FIWARE_SERVICE, originalService);
+        headers.put(CommonConstants.HEADER_FIWARE_SERVICE_PATH, originalServicePath);
+        headers.put(NGSIConstants.FLUME_HEADER_MAPPED_SERVICE, mappedService);
+        headers.put(NGSIConstants.FLUME_HEADER_MAPPED_SERVICE_PATH, mappedServicePath);
+        //NGSICKANSink.ColumnAggregator columnAggregator = ngsickanSink.new ColumnAggregator(false, false, false, false);
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Leftover comment? Should be removed?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 4]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Leftover comment? Should be removed?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 5]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Leftover comment? Should be removed?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 6]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
+        if (columnAggregator.getAggregation().equals(correctAggregation)) {
+            assertTrue(true);
+        } else {
+            fail();
+            System.out.println(getTestTraceHead("[NGSIHDFSSink.testNativeTypeColumnCsv]")
+                    + "-  FAIL");
+        }
+
+    }//testNativeTypeColumnCsv
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]:
```suggestion
    } // testNativeTypeColumnCsv
```

To keep same style than existing ones

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 7]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
+                "{\"recvTimeTs\":\"1461136795\",\"recvTime\":\"2016-04-20T07:19:55.801Z\",\"fiwareServicePath\":\"somePath\",\"entityId\":\"someId\",\"entityType\":\"someType\",\"attrName\":\"someString2\",\"attrType\":\"string\",\"attrValue\":\"\",\"attrMd\":[]}";
+        if (rowAggregator.getAggregation().equals(correctAggregation)) {
+            assertTrue(true);
+        } else {
+            fail();
+            System.out.println(getTestTraceHead("[NGSIHDFSSink.testNativeTypeRowJson]")
+                    + "-  FAIL");
+        }
+
+    }//testNativeTypeRowCsv
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]:
```suggestion
    } // testNativeTypeRowCsv
```

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 8]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
+        String mappedServicePath = "newPath";
+        Map<String, String> headers = new HashMap<>();
+        headers.put(NGSIConstants.FLUME_HEADER_TIMESTAMP, timestamp);
+        headers.put(CommonConstants.HEADER_CORRELATOR_ID, correlatorId);
+        headers.put(NGSIConstants.FLUME_HEADER_TRANSACTION_ID, transactionId);
+        headers.put(CommonConstants.HEADER_FIWARE_SERVICE, originalService);
+        headers.put(CommonConstants.HEADER_FIWARE_SERVICE_PATH, originalServicePath);
+        headers.put(NGSIConstants.FLUME_HEADER_MAPPED_SERVICE, mappedService);
+        headers.put(NGSIConstants.FLUME_HEADER_MAPPED_SERVICE_PATH, mappedServicePath);
+        //NGSICKANSink.ColumnAggregator columnAggregator = ngsickanSink.new ColumnAggregator(false, false, false, false);
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Leftover comment? Should be removed?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 9]
```diff
@@ -1196,5 +1212,352 @@ private Context createContext(String backendImpl, String backendMaxConns, String
...
+                "1461136795,2016-04-20T07:19:55.801Z,somePath,someId,someType,someString2,string,,hdfs:///user/null/someServicesomePath/someId_someTypexffffsomeString2xffffstring/someId_someTypexffffsomeString2xffffstring.txt";
+        if (rowAggregator.getAggregation().equals(correctAggregation)) {
+            assertTrue(true);
+        } else {
+            fail();
+            System.out.println(getTestTraceHead("[NGSIHDFSSink.testNativeTypeRowCsv]")
+                    + "-  FAIL");
+        }
+
+    }//testNativeTypeRowCsv
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]:
```suggestion
    } // testNativeTypeRowCsv
```

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java (hunk 1 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: Fixed in [c434ab4](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/c434ab4ded7d5c6ba56544524927822c68aeaa4b)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 3 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 4 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 5 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 6 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 7 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 8 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 9 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [9e61357](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/9e61357a49559bffebc01a71389c87be49ebd72c)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java (hunk 2 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: Fixed in [d405b06](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/d405b0669fc29cd264e7e161236606bc3b8b75b9)

File: CHANGES_NEXT_RELEASE [hunk 10]
```diff
@@ -8,3 +8,4 @@ Fix: Changing cygnus port does not work (#1698)
 [cygnus-ngsi] Changed log level for some events (#1769)
 [cygnus-ngsi][PostgisSink, PostGRESQLSink] Support for native Postgres and Postgis types (#1780)
 [cygnus-ngsi][MySQL] Support for native MySQL types (#1780)
+[cygnus-ngsi][HDFS] Encode value when it cotains commas (#1785)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Typo

```suggestion
[cygnus-ngsi][HDFS] Encode value when it contains commas (#1785)
```

Comment (IvanHdzC) [reply CHANGES_NEXT_RELEASE]: Thks. Fixed in [0a63bae](https://github.com/telefonicaid/fiware-cygnus/pull/1799/commits/0a63bae73f14a66383707ae6c1ac6111625ad34f)

**Comment by manucarrace** [comment]
HDFS sink with csv-column persistence is quite releated with Hadoop Hive. It is possible to create Hive tables with HDFS files as source data. Hive parser doesn't understand a value between "" if it has a comma (e.g. an attribute with value "Hi, how do you do?" will be interpreted as two fields).

PR to be closed by setting ";" as default separator. 
Another PR to be merged adding a recommendation in documentation to use ";" as csv separator character (#1809).

**Pull request closed without merging**